### DISCUSSION TITLE
Design for `mdit-plugin-columns`

### DISCUSSION THREAD

**Discussion by kaikaibenkai** [mdit-plugins/mdit-plugins#37] [Ideas]
# Summary

Provide a container syntax that can evenly split column widths, like [columns in Lark Docs](https://www.larksuite.com/hc/en-US/articles/294802473696-use-columns-in-docs).

# Basic example

([idea](https://github.com/mdit-plugins/mdit-plugins/issues/36#issuecomment-1826821889))

```md
::: columns

@column

Apple

@column

Banana

@column

Orange
:::
```

...renders to:
```html
<div style="width: 100%; display: flex;">
  <p style="flex: 1;">Apple</p>
  <p style="flex: 1;">Banana</p>
  <p style="flex: 1;">Orange</p>
</div>
```

# Motivation

Bloggers often find it convenient to arrange several images in a row. A container with uniform column width would be of great help.
Markdown table struggles with this task due to self-adaption column width and borders. So we can make a plugin `mdit-plugin-columns`.

# Detailed design

Syntax:
```md
::: columns

@column

Content in column 1...

@column

Content in column 2...

...

@column

Content in column n...
:::
```

...renders to:
```html
<div style="width: 100%; display: flex;">
  <p style="flex: 1;">Content in column 1...</p>
  <p style="flex: 1;">Content in column 2...</p>
  ...
  <p style="flex: 1;">Content in column n...</p>
</div>
```

# Alternatives

## Use `width` instead of `flex: 1`

Note: The sum of the widths of the child elements should be equal to 100%, which means that the width of the last element should be subtracted from the allocated width by 100%.

<details>
<summary>Exmaple: 3 columns</summary>

```md
::: columns

@column

Apple

@column

Banana

@column

Orange
:::
```

...renders to:
```html
<div style="width: 100%; display: flex;">
  <p style="width: 33.3%;">Apple</p>
  <p style="width: 33.3%;">Banana</p>
  <p style="width: 33.4%;">Orange</p>
</div>
```
</details>

**Comment by kaikaibenkai** [comment]
- [x] ~~redesign syntax ([idea](https://github.com/mdit-plugins/mdit-plugins/issues/36#issuecomment-1828034995))~~
- [ ] padding between columns

**Reply by Mister-Hope** [reply]
Actually I don't find it helpful to nest the column, in this case the @ directive can be good, since it's simple.

**Reply by Mister-Hope** [reply]
Pointing out the case because I find it impossible to achieve tab nesting,

**Reply by kaikaibenkai** [reply]
Perhaps nesting columns rarely occurs, much like tables

**Reply by kaikaibenkai** [reply]
Do you think

```md
::: columns {4px}
..
:::
```

is a suitable design for setting padding between columns?

**Reply by Mister-Hope** [reply]
Nope, but I.agree supporting attr syntax
